File: apps/desktop/src/lib/fileWatcher.ts
// File Watcher client - Tauri invoke wrappers for file watching
// Monitors workspace files for external changes (outside the app)

import { invoke } from '@tauri-apps/api/core';
import { listen, type UnlistenFn } from '@tauri-apps/api/event';

// ============================================================================
// Types (matching Rust types)
// ============================================================================

export interface FileChangeEvent {
  change_type: 'modify' | 'create' | 'delete';
  file_key: string;
  timestamp: string;
}

// Transformed type for frontend use (camelCase)
export interface FileChange {
  changeType: 'modify' | 'create' | 'delete';
  fileKey: string;
  timestamp: Date;
}

// ============================================================================
// File Watcher Client
// ============================================================================

class FileWatcherClient {
  private unlistenFn: UnlistenFn | null = null;
  private changeCallback: ((change: FileChange) => void) | null = null;

  /**
   * Start watching a workspace for file changes
   */
  async start(workspaceRoot: string): Promise<void> {
    await invoke('file_watcher_start', { workspaceRoot });
  }

  /**
   * Stop watching a workspace
   */
  async stop(workspaceRoot: string): Promise<void> {
    await invoke('file_watcher_stop', { workspaceRoot });
  }

  /**
   * Mark a file as being saved by the app (to ignore the change event)
   * Call this BEFORE writing to a file
   */
  async markSaving(workspaceRoot: string, fileKey: string): Promise<void> {
    await invoke('file_watcher_mark_saving', { workspaceRoot, fileKey });
  }

  /**
   * Clear the saving mark after save completes
   * Call this AFTER writing to a file
   */
  async clearSaving(workspaceRoot: string, fileKey: string): Promise<void> {
    await invoke('file_watcher_clear_saving', { workspaceRoot, fileKey });
  }

  /**
   * Listen for file change events from the backend
   * Returns an unlisten function to stop listening
   */
  async onFileChange(callback: (change: FileChange) => void): Promise<UnlistenFn> {
    // Store callback for internal use
    this.changeCallback = callback;

    // Listen for events from Rust
    this.unlistenFn = await listen<FileChangeEvent>('file-watcher:change', (event) => {
      const payload = event.payload;
      callback({
        changeType: payload.change_type as FileChange['changeType'],
        fileKey: payload.file_key,
        timestamp: new Date(payload.timestamp),
      });
    });

    return this.unlistenFn;
  }

  /**
   * Stop listening for file change events
   */
  unlisten(): void {
    if (this.unlistenFn) {
      this.unlistenFn();
      this.unlistenFn = null;
      this.changeCallback = null;
    }
  }
}

export const fileWatcherClient = new FileWatcherClient();

File: apps/desktop/src-tauri/src/commands/file_watcher.rs
// File watcher commands - IPC handlers for file watching

use crate::services::file_watcher::FileWatcher;
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;
use tauri::Runtime;
use tokio::sync::RwLock;
use tracing::{debug, info};

/// Registry of file watchers (one per workspace)
pub struct FileWatcherRegistry {
    watchers: HashMap<String, Arc<RwLock<FileWatcher>>>,
}

impl FileWatcherRegistry {
    pub fn new() -> Self {
        Self {
            watchers: HashMap::new(),
        }
    }

    pub fn get(&self, workspace_root: &str) -> Option<Arc<RwLock<FileWatcher>>> {
        self.watchers.get(workspace_root).cloned()
    }

    pub fn insert(&mut self, workspace_root: String, watcher: FileWatcher) {
        self.watchers
            .insert(workspace_root, Arc::new(RwLock::new(watcher)));
    }

    pub fn remove(&mut self, workspace_root: &str) -> Option<Arc<RwLock<FileWatcher>>> {
        self.watchers.remove(workspace_root)
    }
}

impl Default for FileWatcherRegistry {
    fn default() -> Self {
        Self::new()
    }
}

/// State for file watchers
pub struct FileWatcherState {
    pub registry: RwLock<FileWatcherRegistry>,
}

impl FileWatcherState {
    pub fn new() -> Self {
        Self {
            registry: RwLock::new(FileWatcherRegistry::new()),
        }
    }
}

impl Default for FileWatcherState {
    fn default() -> Self {
        Self::new()
    }
}

// ============================================================================
// Tauri Commands
// ============================================================================

/// Start watching a workspace for file changes
#[tauri::command]
pub async fn file_watcher_start<R: Runtime>(
    app: tauri::AppHandle<R>,
    state: tauri::State<'_, FileWatcherState>,
    workspace_root: String,
) -> Result<(), String> {
    info!("Starting file watcher for: {}", workspace_root);

    let mut registry = state.registry.write().await;

    // Check if already watching
    if registry.get(&workspace_root).is_some() {
        debug!("File watcher already running for: {}", workspace_root);
        return Ok(());
    }

    // Create and start watcher
    let mut watcher = FileWatcher::new(PathBuf::from(&workspace_root), None);
    watcher.start(app)?;

    registry.insert(workspace_root, watcher);

    Ok(())
}

/// Stop watching a workspace
#[tauri::command]
pub async fn file_watcher_stop<R: Runtime>(
    _app: tauri::AppHandle<R>,
    state: tauri::State<'_, FileWatcherState>,
    workspace_root: String,
) -> Result<(), String> {
    info!("Stopping file watcher for: {}", workspace_root);

    let mut registry = state.registry.write().await;

    if let Some(watcher) = registry.remove(&workspace_root) {
        let mut w = watcher.write().await;
        w.stop();
    }

    Ok(())
}

/// Mark a file as being saved (to ignore the change event)
#[tauri::command]
pub async fn file_watcher_mark_saving<R: Runtime>(
    _app: tauri::AppHandle<R>,
    state: tauri::State<'_, FileWatcherState>,
    workspace_root: String,
    file_key: String,
) -> Result<(), String> {
    debug!("Marking file as saving: {}", file_key);

    let registry = state.registry.read().await;

    if let Some(watcher) = registry.get(&workspace_root) {
        let w = watcher.read().await;
        w.mark_saving(&file_key);
    }

    Ok(())
}

/// Clear the saving mark after save completes
#[tauri::command]
pub async fn file_watcher_clear_saving<R: Runtime>(
    _app: tauri::AppHandle<R>,
    state: tauri::State<'_, FileWatcherState>,
    workspace_root: String,
    file_key: String,
) -> Result<(), String> {
    debug!("Clearing saving mark for: {}", file_key);

    let registry = state.registry.read().await;

    if let Some(watcher) = registry.get(&workspace_root) {
        let w = watcher.read().await;
        w.clear_saving(&file_key);
    }

    Ok(())
}
